services/bots/src/github-webhook/handlers/validate-cla.ts

```diff
@@ -0,0 +1,269 @@
+    await this.checkAuthors(eventData, commits.data);
+  }
+
+  getCommits(eventData: Record<string, any>) {
+    return this.githubApiClient.pulls.listCommits({
+      owner: eventData.repository.owner.login,
+      repo: eventData.repository.name,
+      pull_number: eventData.number,
+      per_page: 100,
+    });
+  }
+
+  async checkAuthors(eventData: Record<string, any>, commits: any[]) {
+    const authorsWithSignedCLA: Set<string> = new Set();
+    const authorsNeedingCLA: { sha: string; login: string }[] = [];
+    const commitsWithoutLogins: { sha: string; maybeText: string; maybeEmailAddress: boolean }[] =
+      [];
+
+    for await (const commit of commits) {
+      if (commit?.author?.type === 'Bot' || ignoredAuthors.has(commit?.commit?.author?.email)) {
+        return;
+      }
+
+      if (!commit?.author && commit?.commit?.author?.email) {
+        const maybeCheck = commit?.commit?.author?.email?.indexOf('@') > -1;
+        commitsWithoutLogins.push({
+          sha: commit.sha,
+          maybeText: maybeCheck
+            ? `This commit has something that looks like an email address (${commit.commit.author.email}). Maybe try linking that to GitHub?.`
+            : 'No email found attached to the commit.',
+          maybeEmailAddress: maybeCheck,
+        });
+      }
+
+      if (!authorsWithSignedCLA.has(commit.author.login)) {
+        const ddbEntry = await this.ddbClient
+          .getItem({
+            Key: { github_username: commit.author.login },
+            TableName: '',
+          })
+          .promise();
+
+        if (ddbEntry.$response.error) {
+          authorsNeedingCLA.push({ login: commit.author.login, sha: commit.sha });
+        } else {
+          authorsWithSignedCLA.add(commit.author.login);
+        }
+      }
+    }
+
+    if (commitsWithoutLogins.length) {
+      await Promise.all([
+        this.githubApiClient.issues.createComment({
+          owner: eventData.repository.owner.login,
+          repo: eventData.repository.name,
+          issue_number: eventData.number,
+          body: noLoginOnShaComment(
+            commitsWithoutLogins,
+            eventData.pull_request.user.login,
+            `https://github.com/${eventData.repository.full_name}/pull/${eventData.number}/commits/`,
+          ),
+        }),
+        ...commitsWithoutLogins.map((commit) =>
+          this.githubApiClient.repos.createCommitStatus({
+            owner: eventData.repository.owner.login,
+            repo: eventData.repository.name,
+            sha: commit.sha,
+            state: 'failure',
+            description: 'Commit(s) are missing a linked GitHub user.',
+            context: 'cla-bot',
+          }),
+        ),
+        this.githubApiClient.issues.addLabels({
+          owner: eventData.repository.owner.login,
+          repo: eventData.repository.name,
+          issue_number: eventData.number,
+          labels: ['cla-error'],
+        }),
+      ]);
+      return;
```

Comment:
Why don't we process `authorsNeedingCLA` anymore?

Reply:
It's done on L144 <https://github.com/home-assistant/service-hub/pull/2/files#diff-5b3cb5674142f1f25e88e28fa5773ab6c3fcfa4edb7737a8309d6f490bf229b6R135>?

Reply:
but not if we don't hit this return ?

Reply:
That is correct, do we want that?
I don't think that's how it does it now

Reply:
If we get here something is wrong with the commit data, wrong email and/or user.
There is no point in "Abc" signing when it was really "Xyz" that did the commit 🤷‍♂️

Reply:
Aren't we analyzing all the different commits? It could happen that commit 1 had a wrong email, commit 2 has a person who hasn't signed the CLA.

Anyway, it's all edge cases, it's fine.